// buttercoin-engine/lib/operations.js
// Generated by CoffeeScript 1.6.3
(function() {
  var create_results, op, _i, _len, _ref;

  _ref = ['ADD_DEPOSIT', 'WITHDRAW_FUNDS', 'CREATE_LIMIT_ORDER', 'CANCEL_ORDER', 'GET_BALANCES', 'OPEN_ORDERS', 'ORDER_INFO', 'BALANCES', 'TICKER', 'SEND_BITCOINS'];
  for (_i = 0, _len = _ref.length; _i < _len; _i++) {
    op = _ref[_i];
    module.exports[op] = op;
  }

  op = module.exports;

  create_results = {
    ADD_DEPOSIT: function() {},
    BALANCES: function(account_id, balances) {
      return {
        result: op.BALANCES,
        account: account_id,
        balances: balances
      };
    },
    OPEN_ORDERS: function(account_id, orders) {
      return {
        result: op.OPEN_ORDERS,
        account: account_id,
        orders: orders
      };
    },
    TICKER: function(currency, bid, ask, last) {
      return {
        result: op.TICKER,
        currency: currency,
        bid: bid,
        ask: ask,
        last: last
      };
    }
  };

  module.exports.create = {
    result: create_results
  };

}).call(this);
